Allow users to run crawls with 1 or 2 browser windows

tw4l opened the pull request:
Fixes #2425 

## Changed

- Switch backend to primarily using number of browser windows rather than scale multiplier (including migration to calculate `browserWindows` from `scale` for existing workflows and crawls)
- Still support `scale` in addition to `browserWindows` in input models for creating and updating workflows and re-adjusting live crawl scale for backwards compatibility
- Adds new `max_browser_windows` value to Helm chart, but calculates the value from `max_crawl_scale` as fallback for users with that value already set in local charts
- Rework frontend to allow users to select multiples of `crawler_browser_instances` or any value below `crawler_browser_instances` for browser windows. For instance, with `crawler_browser_instances=4` and `max_browser_windows=8`, the user would be presented with the following options: 1, 2, 3, 4, 8
- Sets maximum width of screencast to image width returned by `message`

tw4l committed:
Modify backend scale to be number of browser windows

tw4l committed:
Fix SettingsReponse model

tw4l committed:
Update frontend for simplified maxBrowserWindows

tw4l committed:
Import math

tw4l committed:
Operator fixups

tw4l committed:
Handle case where scale < workers per pod

tw4l committed:
Add pylint comment

tw4l committed:
Update API settings test

tw4l committed:
Fix pylint comment

tw4l committed:
Use crawl.scale, add lots of debug print logging

tw4l committed:
Fixups

tw4l committed:
Consolidate print logging lines

tw4l committed:
Remove some debug print logging

tw4l committed:
Temp: Debug print log sync_crawls return

ikreymer committed:
rebase fix

ikreymer committed:
work

ikreymer committed:
switch back to last pod

ikreymer committed:
fix remainder check

ikreymer committed:
rename priorities to use max_browser_windows

ikreymer committed:
cleanup, use desired pod count with page check to avoid starting up more pods than pages

tw4l committed:
Fix screencast window count

tw4l committed:
Rename scale fields to distinguish pods from browser windows

tw4l committed:
Fix linting

tw4l committed:
Undo change to worker index calculation for screenshots

tw4l committed:
Remove unused variable

tw4l committed:
Add separate browserWindows on backend alongside scale

tw4l committed:
Update frontend to use browserWindows not scale

tw4l committed:
Fix types

tw4l committed:
Calculate browserWindows as needed on org import

tw4l committed:
Fix backend typing

tw4l committed:
Fix frontend compilation errors

tw4l committed:
More backend fixups

tw4l committed:
Update CrawlSpec model to have both scale and browser windows

tw4l committed:
Add browser windows to crawl_job

ikreymer committed:
compute resources for remainder browsers

tw4l committed:
Keep default of scale=1 if browser windows/scale not specified

tw4l committed:
Fix broken org import test

ikreymer committed:
additional fixes for N browses:
- ensure correct adjustment on frontend for unequal browser amounts
- ensure pods correctly restarted if number of browser for each pod changes, track lastWorkers per pod
- ensure if both max browsers and max scale is specified, max browsers if chosen, compute max scale from max browsers,
otherwise compute max browsers from max scale

ikreymer committed:
ensure backwards compatible with max_crawl_scale if no max_browser_windows

ikreymer committed:
fix tests

ikreymer committed:
switch browser windows to text box

tw4l committed:
Calculate scale at time of need instead of storing in db

This will keep scale and browserWindows from getting out of sync
when the number of workers per pod on an instance changes.

tw4l committed:
Update org import for change

tw4l committed:
More fixups for removing scale from db

tw4l committed:
Add migration to convert scale to browserWindows in db

Comment out unsetting of scale for now for easier testing on dev

ikreymer committed:
frontend: custom range for browser windows, by 1 until browser instances,
then in multiples through max

tw4l committed:
Don't unset scale in migration

tw4l committed:
Store scale in crawl object

tw4l committed:
Update scale in crawl model when crawl is live rescaled

tw4l committed:
Add some tests

tw4l committed:
Update expected totals in tests

tw4l commented on backend/btrixcloud/operator/crawls.py:
```diff
@@ -359,9 +365,27 @@ async def sync_crawls(self, data: MCSyncData):
-        for i in range(0, status.scale):
+        print(f"status.scale: {status.scale}", flush=True)
+        print(f"crawl.browser_windows: {crawl.browser_windows}", flush=True)
+
```
```suggestion
```

To remove, just for debugging

tw4l commented on backend/btrixcloud/operator/crawls.py:
```diff
@@ -482,24 +515,57 @@ def _load_crawler(self, params, i, status: CrawlStatus, children, paused: bool):
+            f"crawler pod {i}, {worker_count} workers",
+            flush=True,
+        )
+
```
```suggestion
```

To remove, just for debugging

tw4l commented on backend/btrixcloud/operator/crawls.py:
```diff
@@ -521,9 +587,9 @@ def _qa_configmap_update_needed(self, name, configmap):
-        desired_scale: int,
+        crawl: CrawlSpec,
         redis: Redis,
+        # pylint: disable=unused-argument
```
```suggestion
```

tw4l committed:
Remove outdated pylint comment

SuaYoo committed:
set max width

SuaYoo commented:
@tw4l Updated to set the maximum width to the screencast image width:

<img width="977" alt="Screenshot 2025-05-28 at 7 06 44 PM" src="https://github.com/user-attachments/assets/0c1484d6-c993-4def-aace-261af67c5b60" />

SuaYoo committed:
fix spinner

ikreymer committed:
cleanup:
- rename pod_count -> scale for consistency
- remove debug logging
- simplify update_scale to remove cast

ikreymer committed:
Update backend/btrixcloud/operator/crawls.py

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

SuaYoo reviewed (commented):
Looks good testing locally, if it's easy to do we should add a deprecation notice for the `scale` API field.

tw4l committed:
Add deprecated flag to Scale

ikreymer reviewed (approved):
Looks good! Nice work, tested on dev with different number of windows!
